### PR TITLE
Fixes BXC #53775

### PR DESCRIPTION
**Pull request by Jammer77**
Implemented focus for EventBoxTooltip

### CODE REVIEW

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Components/EventBoxTooltip.cs [hunk 1]
```diff
@@ -43,24 +45,47 @@ public class EventBoxTooltip : IDisposable
...
+
 		/// <summary>
 		/// The EventBox should have Visible set to false otherwise the tooltip pop window
 		/// will have the wrong location.
 		/// </summary>
+
+		ImageView image;
+		Pixbuf normalPixbuf;
+		Pixbuf activePixbuf;
+
```
Comment (mrward) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Components/EventBoxTooltip.cs]: This has moved the comment away from the constructor. I would have put these new fields with the eventBox field at the top of the class.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Components/EventBoxTooltip.cs [hunk 2]
```diff
@@ -146,6 +171,39 @@ public void Dispose ()
 
 		public TaskSeverity? Severity { get; set; }
 		public PopupPosition Position { get; set; }
-	}
+	}
+#region Extension method for Pixbuf
```
Comment (mrward) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Components/EventBoxTooltip.cs]: Not sure we need a region here.

Comment (mrward) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Components/EventBoxTooltip.cs]: If this is a useful class should we extract it into a PixbufExtensions class?

**Comment by iainx** [comment]
Looks good to me once Matt's comments have been fixed

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Components/EventBoxTooltip.cs (hunk 2 above)
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Components/EventBoxTooltip.cs]: I think this should be better suited for ImageService, as doing this means we recreate pixbufs on every switch. ImageService has caching for pixbufs in different variants.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Components/EventBoxTooltip.cs [hunk 3]
```diff
@@ -146,6 +171,39 @@ public void Dispose ()
...
+	internal static class PixbufExtension
+	{
+		public static unsafe Pixbuf ColorShiftPixbuf (this Pixbuf src, byte shift = 120)
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Components/EventBoxTooltip.cs]: The whole method doesn't need to be unsafe, you can use an unsafe block for the math code.

**Comment by slluis** [comment]
@iainx @Jammer77 This fix may work, but it is not entirely correct. Conversion from Xwt.Image to Pixbuf should be avoided unless there is really no other option. The problem is that a Pixbuf is a bitmap, while an Xwt.Image may not be a bitmap, so when doing the conversion, some information may be lost. Most of our icons have low and high resolution variants. ImageView handles variant selection automatically, so that when the image is rendered in a retina display, the high res version is used. If the image is converted to a pixbuf and then converted back to an Xwt.Image, this capability is lost.

A solution would be to draw the focus in the image using Xwt.ImageBuilder. It would not be possible to do color shifting, but we can do something else to show the focus.

**Comment by slluis** [comment]
In any case, ColorShiftPixbuf should not be public.

**Pull request merged**